NullPointerException at test with timeOut

### TestNG Version
7.1.0

### Expected behavior
no error

### Actual behavior
```
1 test completed, 1 failed
> Task :backend:server:test FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':backend:server:test'.
> Failed to notify test listener internal.
   > java.lang.NullPointerException (no error message)
   > Failed to notify test listener.
      > java.lang.NullPointerException (no error message)
      > java.lang.NullPointerException (no error message)
* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':backend:server:test'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:205)
	at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:203)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:184)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:114)
	at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
	at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: org.gradle.internal.event.ListenerNotificationException: Failed to notify test listener internal.
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:86)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:325)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:235)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:141)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy147.completed(Unknown Source)
	at org.gradle.api.internal.tasks.testing.results.StateTrackingTestResultProcessor.completed(StateTrackingTestResultProcessor.java:96)
	at org.gradle.api.internal.tasks.testing.results.AttachParentTestResultProcessor.completed(AttachParentTestResultProcessor.java:56)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.FailureHandlingDispatch.dispatch(FailureHandlingDispatch.java:30)
	at org.gradle.internal.dispatch.AsyncDispatch.dispatchMessages(AsyncDispatch.java:87)
	at org.gradle.internal.dispatch.AsyncDispatch.access$000(AsyncDispatch.java:36)
	at org.gradle.internal.dispatch.AsyncDispatch$1.run(AsyncDispatch.java:71)
	at org.gradle.internal.concurrent.InterruptibleRunnable.run(InterruptibleRunnable.java:42)
	at org.gradle.internal.operations.CurrentBuildOperationPreservingRunnable.run(CurrentBuildOperationPreservingRunnable.java:42)
	... 3 more
Caused by: java.lang.NullPointerException
	at org.gradle.internal.build.event.types.DefaultFailure.fromThrowable(DefaultFailure.java:56)
	at org.gradle.tooling.internal.provider.runner.ClientForwardingTestOperationListener.convertExceptions(ClientForwardingTestOperationListener.java:159)
	at org.gradle.tooling.internal.provider.runner.ClientForwardingTestOperationListener.adapt(ClientForwardingTestOperationListener.java:150)
	at org.gradle.tooling.internal.provider.runner.ClientForwardingTestOperationListener.finished(ClientForwardingTestOperationListener.java:88)
	at org.gradle.internal.operations.DefaultBuildOperationListenerManager$ProgressShieldingBuildOperationListener.finished(DefaultBuildOperationListenerManager.java:128)
	at org.gradle.internal.operations.DefaultBuildOperationListenerManager$1.finished(DefaultBuildOperationListenerManager.java:61)
	at org.gradle.api.internal.tasks.testing.operations.TestListenerBuildOperationAdapter.completed(TestListenerBuildOperationAdapter.java:69)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:382)
	at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:364)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:352)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:339)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:352)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:339)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:231)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:150)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
	... 22 more
* Get more help at https://help.gradle.org
Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.2.1/userguide/command_line_interface.html#sec:command_line_warnings
BUILD FAILED in 1m 34s

```

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [X] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans


### Test case sample
gradle test config:
```
test {
    maxHeapSize = "4g"
    jvmArgs += ["-javaagent:" + aspectjWeaver.path]
    useTestNG() {
        useDefaultListeners = true
        suiteXmlFiles = [
                file("src/test/resources/suite.xml")
        ]
    }
}
```
suite.xml:
```
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >

<suite name="UnitTestSuite" verbose="1">
    <test name="UnitTest">
        <groups>
            <run>
                <exclude name="verylong"/>
                <exclude name="long"/>
                <exclude name="manual"/>
                <exclude name="network"/>
            </run>
        </groups>
        <packages>
            <package name="basepackage.*"/>
        </packages>
    </test>
</suite>
```

Java spring boot test with timeout fails `@Test(timeOut = 10000)`, but without timeOut works.
Note that timeout is never reached, setting timeOut to a very large value results in same error.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-2251: NullPointerException at test with timeOut (Krishnan Mahadevan)
 Fixed: GITHUB-2249: Not abstract super-classes mess up test run order (Sergii Kim)
 Fixed: GITHUB-2195: NPE Using groups and @Before/@AfterMethod with alwaysRun and dependsOnMethods (Tomas & Julien Herr)
 Fixed: GITHUB-2238: Parameter values should be overridable from JVM arguments (Krishnan Mahadevan)

--- a/src/main/java/org/testng/internal/InvokeMethodRunnable.java
+++ b/src/main/java/org/testng/internal/InvokeMethodRunnable.java
@@ -50,7 +50,11 @@ public class InvokeMethodRunnable implements Callable<Void> {
               m_instance, m_parameters, m_hookable, m.getMethod(), m_testResult);
         }
       } catch (Throwable e) {
-        t = new TestNGRuntimeException(e.getCause());
+        Throwable cause = e.getCause();
+        if (cause == null) {
+          cause = e;
+        }
+        t = new TestNGRuntimeException(cause);
       }
       if (null != t) {
         Thread.currentThread().interrupt();

--- a/src/test/java/test/hook/issue2251/AbstractBaseTestCase.java
+++ b/src/test/java/test/hook/issue2251/AbstractBaseTestCase.java
@@ -1,0 +1,42 @@
+package test.hook.issue2251;
+
+import java.lang.reflect.InvocationTargetException;
+import org.testng.IHookCallBack;
+import org.testng.IHookable;
+import org.testng.ITestResult;
+
+/**
+ * This test class apes on a bare essential What Spring TestNG provides as a base class in terms of
+ * running Spring based tests. The following methods have been duplicated from
+ * org.springframework.test.context.testng.AbstractTestNGSpringContextTests to simulate the bug. 1.
+ * throwAsUncheckedException() 2. getTestResultException() 3. throwAs()
+ */
+public class AbstractBaseTestCase implements IHookable {
+
+  @Override
+  public void run(IHookCallBack callBack, ITestResult testResult) {
+    callBack.runTestMethod(testResult);
+    Throwable t = getTestResultException(testResult);
+    if (t != null) {
+      throwAsUncheckedException(t);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T extends Throwable> void throwAs(Throwable t) throws T {
+    throw (T) t;
+  }
+
+  private void throwAsUncheckedException(Throwable t) {
+    throwAs(t);
+  }
+
+
+  private Throwable getTestResultException(ITestResult testResult) {
+    Throwable testResultException = testResult.getThrowable();
+    if (testResultException instanceof InvocationTargetException) {
+      testResultException = testResultException.getCause();
+    }
+    return testResultException;
+  }
+}

--- a/src/test/java/test/hook/issue2251/IssueTest.java
+++ b/src/test/java/test/hook/issue2251/IssueTest.java
@@ -1,0 +1,21 @@
+package test.hook.issue2251;
+
+import org.assertj.core.api.Assertions;
+import org.testng.TestListenerAdapter;
+import org.testng.TestNG;
+import org.testng.annotations.Test;
+import test.SimpleBaseTest;
+
+public class IssueTest extends SimpleBaseTest {
+
+  @Test(description = "GITHUB-2251")
+  public void runTest() {
+    TestNG testng = create(SampleTestCase.class);
+    TestListenerAdapter l = new TestListenerAdapter();
+    testng.addListener(l);
+    testng.run();
+    Throwable t = l.getFailedTests().get(0).getThrowable();
+    Assertions.assertThat(t).isInstanceOf(NullPointerException.class);
+  }
+
+}

--- a/src/test/java/test/hook/issue2251/SampleTestCase.java
+++ b/src/test/java/test/hook/issue2251/SampleTestCase.java
@@ -1,0 +1,19 @@
+package test.hook.issue2251;
+
+import org.testng.annotations.Test;
+
+public class SampleTestCase extends AbstractBaseTestCase {
+
+  static class NullExObj {
+
+    @Override
+    public String toString() {
+      throw new NullPointerException("expected");
+    }
+  }
+
+  @Test(timeOut = 1000000) //removing timeout fixes error output
+  public void testError() throws Exception {
+    new NullExObj().toString();
+  }
+}

--- a/src/test/resources/testng.xml
+++ b/src/test/resources/testng.xml
@@ -627,6 +627,7 @@
   <test name="Hookable" >
     <classes>
       <class name="test.hook.HookableTest"/>
+      <class name="test.hook.issue2251.IssueTest"/>
     </classes>
   </test>
 

